subsys/nfc/t4t/isodep.c

```diff
@@ -0,0 +1,875 @@
+						 t4t_isodep.tx_data.len,
+						 ISODEP_FWT_DEACTIVATION);
+		}
+
+		return;
+
+	default:
+		break;
+	}
+
+	err = timeout ? -NFC_T4T_ISODEP_TIMEOUT_ERROR : -NFC_T4T_ISODEP_TRANSMISSION_ERROR;
+
+	if (err) {
+		atomic_set(&t4t_isodep.state, ISODEP_STATE_INITIALIZED);
+		err_notify(err);
+	}
+}
+
+static void isodep_delay_handler(struct k_work *work)
+{
+	isodep_chunk_send();
+}
+
+int nfc_t4t_isodep_init(u8_t *tx_buf, size_t size_tx,
+			u8_t *rx_buf, size_t size_rx,
+			const struct nfc_t4t_isodep_cb *cb)
+{
+	if (!tx_buf || !rx_buf) {
+		return -EINVAL;
+	}
+
+	if (size_tx < T4T_FSD_MIN) {
+		LOG_ERR("Tx buffer is to small. Minimum Tx buffer size %d\n",
+			T4T_FSD_MIN);
+		return -EINVAL;
+	}
+
+	if (size_rx < T4T_FSD_MIN) {
+		LOG_ERR("Rx buffer is to small. Minimum Rx buffer size %d\n",
+			T4T_FSD_MIN);
+		return -EINVAL;
+	}
+
+	if (atomic_get(&t4t_isodep.state) != ISODEP_STATE_UNINITIALIZED) {
+		LOG_ERR("Already initialized");
+
+		return -EACCES;
+	}
+
+	t4t_isodep.tx_data.data = tx_buf;
+	t4t_isodep.tx_data.buf_size = size_tx;
+	t4t_isodep.rx_data.data = rx_buf;
+	t4t_isodep.rx_data.buf_size = size_rx;
+
+	t4t_isodep_cb = cb;
+
+	/** Initialize work queue to handle delay before sending first
+	 * after receipt RATS Response.
+	 */
+	k_delayed_work_init(&isodep_work, isodep_delay_handler);
+
+	atomic_set(&t4t_isodep.state, ISODEP_STATE_INITIALIZED);
+
+	return 0;
+}
+
+void nfc_t4t_isodep_on_timeout(void)
+{
+	isodep_error_handle(true);
+}
+
+int nfc_t4t_isodep_rats_send(enum nfc_t4t_isodep_fsd fsd, u8_t did)
+{
+	u8_t param;
+	u32_t state = atomic_get(&t4t_isodep.state);
+
+	if ((state == ISODEP_STATE_UNINITIALIZED) ||
+	    (state == ISODEP_STATE_TRANSFER)) {
+		return -EACCES;
+	}
+
+	atomic_set(&t4t_isodep.state, ISODEP_STATE_TRANSFER);
+
+	if (did > T4T_DID_MAX) {
+		LOG_ERR("Invalid DID value. It should be between 0-14.");
+
+		return -EINVAL;
+	}
+
+	if (t4t_isodep.tx_data.buf_size < fsd_value_map[fsd]) {
+		LOG_ERR("Invalid FSD value. Increase Tx buffer size or decrease FSD");
+
+		return -ENOMEM;
+	}
+
+	/* Set DID field. */
+	param = did & T4T_RATS_DID_MASK;
+
+	/* Set FSDI field. */
+	param = (fsd << T4T_RATS_FSDI) & T4T_RATS_FSDI_MASK;
+
+	t4t_isodep.tx_data.data[0] = T4T_RATS_CMD;
+	t4t_isodep.tx_data.data[1] = param;
+
+	t4t_isodep.tx_data.len = T4T_RATS_CMD_LEN;
+
+	/* Set reader/writer parameters. */
+	t4t_isodep.tag.did = did;
+	t4t_isodep.fsd = fsd_value_map[fsd];
+
+	t4t_isodep.ats_expected = true;
+
+	t4t_isodep.err_status.last_frame = ISODEP_FRAME_RATS;
+
+	if (t4t_isodep_cb->data_send) {
+		t4t_isodep_cb->data_send(t4t_isodep.tx_data.data,
+					 t4t_isodep.tx_data.len,
+					 T4T_RATS_FDT);
+	}
+
+	return 0;
+}
+
+int nfc_t4t_isodep_tag_deselect(void)
+{
+	size_t index = 0;
+	u8_t *tx_data;
+
+	if (!atomic_cas(&t4t_isodep.state,
+			ISODEP_STATE_SELECTED,
+			ISODEP_STATE_TRANSFER)) {
+		return -EACCES;
+	}
+
+	atomic_set(&t4t_isodep.state, ISODEP_STATE_TRANSFER);
```

Comment:
Atomic CAS should already set the correct state, so this is unnecessary

Reply:
Yes right.